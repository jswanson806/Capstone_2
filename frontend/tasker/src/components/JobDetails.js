import React from 'react';

const JobDetails = ({job}) => {
    return(
        <>
        <h1>{job.title}</h1>
        <p>{job.body}</p>
        </>
    )
}

export default JobDetails;