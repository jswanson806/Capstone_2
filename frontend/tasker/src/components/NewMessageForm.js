import React, {useState} from "react";
import TaskerApi from "../api";

const NewMessageForm = ({convoId, assignedUser, currUser, jobId, onAction, onMessageSent}) => {
    
    const INITIAL_STATE = {
        body: '', 
        sent_by: '', 
        sent_to: '', 
        job_id: ''
    }

    const [formData, setFormData] = useState(INITIAL_STATE);


    const handleSubmit = async (e) => {
        e.preventDefault();
        
        let message = {message: {
            body: formData.body,
            sent_by: currUser.id, 
            sent_to: assignedUser.id, 
            job_id: jobId
        }}
        
        await TaskerApi.createMessage(message);

        if(onAction){
            handleClose(); 
        } else {
            onMessageSent(assignedUser, currUser.id, convoId);
        }

        setFormData(INITIAL_STATE);
    };


    const handleChange = (e) => {
        const {name, value} = e.target;
        setFormData(formData => ({
            ...formData,
            [name]: value
        }))
    }

    const handleClose = () => {
        onAction();
    }
    

    return (
        <div className="convo-new-message-form">
            <form onSubmit={handleSubmit}>
                <label htmlFor="message">Message: </label>
                <input 
                    id="body"
                    type="text"
                    name="body"
                    placeholder="What's on your mind?"
                    data-testid="convo-message-input"
                    value={formData.body}
                    onChange={handleChange}
                />
                <button type="submit" data-testid="convo-form-button">Send</button>
            </form>
            {onAction && (<button onClick={handleClose}>Close</button>)}
        </div>
    )
}

export default NewMessageForm;