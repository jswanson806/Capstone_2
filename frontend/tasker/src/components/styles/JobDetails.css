.carousel {
    width: 100%;
    max-height: 80vh; /* Optional: Limit the maximum height of the Carousel */
    margin: 0 auto;
  }
  
.carousel-item {
  text-align: center; /* Center the images horizontally */
}

.carousel-inner {
  max-height: 100%;
}

.carousel img {
  max-width: 50%;
  max-height: 50%;
  margin: 0 auto;
}

.carousel-item-content {
  position: relative; /* Ensure relative positioning for the parent container */
}

.carousel-caption {
  color: black !important; /* Optional: Set the text color */
  padding: 10px !important; /* Optional: Add padding to the caption */
}

.carousel-control-next-icon, .carousel-control-prev-icon {
  background-color: black !important;
  border: 5px solid black;
  border-radius: 5px;
  padding: 10px;
}

.carousel-indicators li {
  border-top: unset;
  width: 10px;
  border-bottom: unset;
  border: 1px solid transparent;
  border-left-width: 0;
  border-right-width: 0;
  border-radius: 10px;
  background-color: rgba(169, 29, 42, 0.4);
  height: 10px;
  width: 20px;}                                                  
                                                                 
  .carousel-indicators .active {
  border-top: unset;
  border-bottom: unset;
  border: 1px solid transparent;
  border-left-width: 0;
  border-right-width: 0;
  border-radius: 10px;
  background-color: #A91D2A;
  height: 10px;
  width: 50px}

.jobDetails-container {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 2em 0 2em;
}

.jobDetails-centered-card {
  text-align: center;
  max-height: 700px;
  width: 100%;
  margin: 0 auto;
  background-color: #fff;
  border-radius: 20px;
  
  padding: 20px;
  transition: max-height 0.2s ease-in-out;
}

.button {
    width: 20vw;
}

.jobDetails-text {
  text-align: left;
  padding: 10px;
  margin: 10px 5px;
  background-color: #e2e2e2;
}

.card-body {
  border-radius: 10px !important;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1) !important;
  margin-bottom: 50px !important;
}

li {
  color: blue;
  cursor: pointer;
}

.modal-footer {
  justify-content: left !important;
}
