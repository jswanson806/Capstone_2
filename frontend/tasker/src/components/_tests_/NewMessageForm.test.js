import React from 'react';
import {act, render, screen, fireEvent, waitFor} from '@testing-library/react';
import TaskerApi from '../../api.js';
import NewMessageForm from '../NewMessageForm.js';
import JobDetails from '../JobDetails.js';

// NewMessage component props
const currUser = {id: 1, email: "test@email.com", isWorker: true};
const convoId = 'u1u2j1';
const onAction = jest.fn();
const onClose = jest.fn();
const onMessageSent = jest.fn();
const assignedUser = {id: 1};
const handleClose = jest.fn();

// JobDetails component props
const job4 = {
    id: 4,
    title: "test job 4",
    body: "job 4 body for testing the job card",
    posted_by: 2,
    status: 'in progress',
    assigned_to: 3,
    start_time: 'test start time',
    payment_due: null,
    address: '123 Test Street',
    before_image_url: 'http://beforeImgUrl2.com',
    after_image_url: null,
    applicants: [1, 2]
};
const userValue3 = { user: {id: 3, email: "test@email.com", isWorker: true, applications: []} };
const onEndWork = jest.fn();
const startWork = jest.fn();
const withdrawApplication = jest.fn();
const onJobComplete = jest.fn();
const toggleDetails = jest.fn();
const triggerEffect = jest.fn();
const fetchCurrUser = jest.fn();

const applications = new Set([job4.id]);

// api mocks
const getBeforeImage = jest.spyOn(TaskerApi, 'getBeforeImage');
const applyToJob = jest.spyOn(TaskerApi, 'applyToJob');
const createMessage = jest.spyOn(TaskerApi, 'createMessage');
const getSingleUser = jest.spyOn(TaskerApi, 'getSingleUser');


describe("login form smoke and snapshot tests", () => {

    it("should render without crashing", async () => {

        render(
            <NewMessageForm
                convoId={convoId}
                currUser={currUser}
                assignedUser={{id: 2}}
                jobId={1}
                onAction={onAction}
                onMessageSent={onMessageSent}
                onClose={onClose}
            />
        );
        
    });

    it("should match snapshot", async() => {

        const {asFragment} = render(
            <NewMessageForm
                convoId={convoId}
                currUser={currUser}
                assignedUser={{id: 2}} 
                jobId={1}
                onAction={onAction}
                onMessageSent={onMessageSent}
                onClose={onClose}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    })
})

describe("Handles form input correctly", () => {

    it('updates form input correctly', async () => {

        render(
            <NewMessageForm
                convoId={convoId}
                currUser={currUser}
                assignedUser={{id: 2}} 
                jobId={1}
                onAction={onAction}
                onMessageSent={onMessageSent}
                onClose={onClose}
            />
        );

        const msgInput = screen.getByTestId("convo-message-input");
        fireEvent.change(msgInput, {target: {value: "test message"}});

        expect(msgInput).toHaveValue("test message");

    })

    it('should handle form submission', async () => {

        render(
            <NewMessageForm
                convoId={convoId}
                currUser={currUser}
                assignedUser={{id: 2}} 
                jobId={1}
                onMessageSent={onMessageSent}
                onClose={onClose}
            />
        );

        const msgInput = screen.getByTestId("convo-message-input");
        fireEvent.change(msgInput, {target: {value: "test message"}});

        const btn = screen.getByTestId("convo-form-button");

        await act(async () => {
            fireEvent.click(btn);
        })

        const message = {message: {
            body: "test message",
            sent_by: currUser.id, 
            sent_to: 2, 
            job_id: 1
        }};
        expect(createMessage).toHaveBeenCalledTimes(1);
        expect(createMessage).toHaveBeenCalledWith(message);

    })

})

describe("Buttons call functions correctly", () => {

    it('sending message calls onMessageSent with correct data', async () => {
        expect.assertions(2)

        render(
            <NewMessageForm
                convoId={convoId}
                currUser={currUser}
                assignedUser={assignedUser}
                onMessageSent={onMessageSent}
                jobId={1}
                onClose={onClose}
            />
        );

        const msgInput = screen.getByTestId("convo-message-input");
        fireEvent.change(msgInput, {target: {value: "test message"}});

        const btn = screen.getByTestId("convo-form-button");

        await act(async () => {
            fireEvent.click(btn);
        })

        expect(onMessageSent).toHaveBeenCalledTimes(1);
        expect(onMessageSent).toHaveBeenCalledWith(assignedUser, currUser.id, convoId);
        
    })

    it('sending message calls onAction when passed from JobDetails component', async () => {
        // expect.assertions(2)
 
        getSingleUser.mockResolvedValueOnce({user: 
            {
                id: 3, 
                firstName: "Joe", 
                lastName: "Doe", 
                email: "test@email.comm", 
                isWorker: true, 
                applications: [1, 2]
            }});

        getBeforeImage.mockResolvedValueOnce({preSignedUrl: 'presigned_before'});

        await act (async () => {
            render(
                <JobDetails
                    user={userValue3.user} 
                    applications={applications} 
                    applyToJob={applyToJob}
                    job={job4} 
                    fetchCurrUser={fetchCurrUser} 
                    onEndWork={onEndWork} 
                    startWork={startWork} 
                    withdrawApplication={withdrawApplication} 
                    onJobComplete={onJobComplete} 
                    toggleDetails={toggleDetails}
                    triggerEffect={triggerEffect}
                />
            );
        }); 

        const button = screen.queryByTestId('jobDetails-message-button');
        expect(button).toBeInTheDocument();
        fireEvent.click(button);
        
        const msgInput = screen.getByTestId("convo-message-input");
        expect(msgInput).toBeInTheDocument();

        fireEvent.change(msgInput, {target: {value: "test message"}});

        const submitBtn = screen.getByTestId("convo-form-button");
        expect(submitBtn).toBeInTheDocument();

        await act(async () => {
            fireEvent.click(submitBtn);
        })

        expect(createMessage).toHaveBeenCalledTimes(1);
        // onAction should have been called, which will hide the NewMessageForm
        expect(msgInput).not.toBeInTheDocument();
        
    })

    it('close calls onClose', async () => {
        expect.assertions(2)

        render(
            <NewMessageForm
                convoId={convoId}
                currUser={currUser}
                assignedUser={assignedUser}
                onMessageSent={onMessageSent}
                jobId={1}
                onClose={onClose}
            />
        );

        const closeBtn = screen.getByTestId("convo-close-button");
        expect(closeBtn).toBeInTheDocument();

        await act(async () => {
            fireEvent.click(closeBtn);
        })

        expect(onClose).toHaveBeenCalledTimes(1);
        
    })

})