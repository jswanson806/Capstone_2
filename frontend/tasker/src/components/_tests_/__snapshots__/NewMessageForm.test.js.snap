// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login form smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="new-message-container"
  >
    <div
      class="new-message-centered-card"
    >
      <form
        class=""
      >
        <div
          class="row form-group"
        >
          <label
            class=""
            for="message"
          >
            New Message: 
          </label>
          <textarea
            class="text-input form-control"
            data-testid="convo-message-input"
            id="body"
            name="body"
            placeholder="What's on your mind?"
            rows="1"
            style="resize: none; overflow-y: hidden;"
          />
        </div>
        <button
          class="button btn btn-info"
          data-testid="convo-form-button"
          type="submit"
        >
          Send
        </button>
        <button
          class="button btn btn-danger"
          data-testid="convo-close-button"
          type="button"
        >
          Close
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
